### PR TITLE
Delete index query and Test cases for MetaFileUtil functions

### PR DESCRIPTION
**Pull request by menakaj**
This pull request contains the following modifications. 
1. Delete index query implementation.
   Query structure: 
              delete-index("path-to-index-folder")
   Note:- The execution of delete index is just as same as update-index. So some changes were done to IndexUpdator.java class so that, index deletion can be done within the same class. Several functionalities were broken in to functions to be used with maximum flexibility. 
   The query level test case is also added.
2. JUnit test cases to test MetaFileUtil functions.
   This contains test cases for following.
-  Generating MD5 checksum for XML files.
-  Creating metadata file
-  Validating the file content
-  Detecting XML file changes and validation.
-  Deleting the metadata file.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,6 +2,5 @@
 .settings
 .classpath
 .idea
-*.iml
```
Comment (prestoncarman) [inline .gitignore:5]: Don't remove.

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 2]
```diff
@@ -22,7 +22,7 @@
         <return type="xsext:numeric?"/>
         <runtime type="scalar" class="org.apache.vxquery.runtime.functions.numeric.FnAbsScalarEvaluatorFactory"/>
     </function>
-
+    
     <!-- fn:adjust-date-to-timezone( $arg as xs:date?) as  xs:date? -->
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml:26]: Rebase of master. These spaces have been removed.

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 3]
```diff
@@ -144,36 +144,44 @@
                  class="org.apache.vxquery.runtime.functions.index.IndexUpdaterEvaluatorFactory"/>
     </function>
 
+    <!-- fn:delete-index($indexFolder  as xs:string?) as  node()* -->
+    <function name="fn:delete-index">
+        <param name="index-folder" type="xs:string?"/>
+        <return type="node()*"/>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml]: Return boolean value on whether the function was successful. Also change update-index.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+            @Override
+            protected void evaluate(TaggedValuePointable[] args, IPointable result) throws SystemException {
+                IndexUpdater updater = new IndexUpdater(args, result, stringp, bbis, di, sb, abvs, nodeIdProvider,
+                        abvsFileNode, nodep, nodeId);
+                try {
+                    updater.setup();
+                    updater.deleteAllIndexes();
+                } catch (IOException | NoSuchAlgorithmException e) {
+                    e.printStackTrace();
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java]: Throw exception.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/IndexUpdater.java [hunk 5]
```diff
@@ -22,9 +22,7 @@
 import org.apache.hyracks.dataflow.common.comm.util.ByteBufferInputStream;
 import org.apache.log4j.Level;
 import org.apache.log4j.Logger;
-import org.apache.lucene.index.IndexWriter;
-import org.apache.lucene.index.IndexWriterConfig;
-import org.apache.lucene.index.Term;
+import org.apache.lucene.index.*;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/IndexUpdater.java]: Don't use import *.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/IndexUpdater.java [hunk 6]
```diff
@@ -261,4 +286,40 @@ public void deleteIndexOfDeletedFiles(Set<String> pathsFromMap, Set<String> path
...
+     */
+    public void deleteAllIndexes() throws SystemException {
+        try {
+            indexWriter.deleteAll();
+            indexWriter.commit();
+            indexWriter.close();
+            metaFileUtil.deleteMetaDataFile();
+
+            for (File f : (new File(indexFolder)).listFiles())
+                Files.delete(f.toPath());
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/IndexUpdater.java:316]: Use {}

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/MetaFileUtil.java [hunk 7]
```diff
@@ -106,4 +107,19 @@ public String generateMD5(File file) throws NoSuchAlgorithmException, IOExceptio
...
     }
 
+    /**
+     * Delete the existing Metadata file.
+     * @return True if deleted, false otherwise.
+     */
+    public boolean deleteMetaDataFile() {
+        try {
+            Files.delete(Paths.get(metaFile.getCanonicalPath()));
+            LOGGER.log(Level.DEBUG, "Metadata file deleted!");
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/MetaFileUtil.java]: Add a condition to the log.

File: vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Validate the content of the file.
+     */
+    @Test
+    public void step3_testValidateMetadataFile() throws IOException, ClassNotFoundException {
+        ConcurrentHashMap<String, XmlMetadata> fromFile = metaFileUtil.readMetaFile();
+        Set<String> from = fromFile.keySet();
+        Set<String> initial = initialMap.keySet();
+
+        System.out.println();
```
Comment (prestoncarman) [inline vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java]: Don't print.

File: vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Validate the updated metadata.
+     */
+    @Test
+    public void step6_testVerifyMetadataChange() throws IOException, ClassNotFoundException {
+        ConcurrentHashMap<String, XmlMetadata> fromFile = metaFileUtil.readMetaFile();
+        Set<String> from = fromFile.keySet();
+        Set<String> modified = modifiedMap.keySet();
+
+        System.out.println();
```
Comment (prestoncarman) [inline vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java]: Remove print statement.

Comment (menakaj) [inline vxquery-core/src/test/java/org/apache/vxquery/indexing/MetaFileUtilTest.java]: remove this

File: vxquery-core/src/test/resources/collection/catalog.txt [hunk 10]
```diff
@@ -0,0 +1,43 @@
+<?xml version="1.0" standalone="yes"?>
+<!--
```
Comment (prestoncarman) [inline vxquery-core/src/test/resources/collection/catalog.txt]: Change .xml?

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java [hunk 11]
```diff
@@ -45,8 +45,8 @@ public IndexDeleteEvaluatorFactory(IScalarEvaluatorFactory[] args) {
         super(args);
     }
 
-    @Override
-    protected IScalarEvaluator createEvaluator(IHyracksTaskContext ctx, IScalarEvaluator[] args) throws AlgebricksException {
+    @Override protected IScalarEvaluator createEvaluator(IHyracksTaskContext ctx, IScalarEvaluator[] args)
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/IndexDeleteEvaluatorFactory.java]: I am not sure why this changed. Our formatting should keep @Override on a separate line.

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/MetaFileUtil.java [hunk 12]
```diff
@@ -59,36 +63,52 @@ public boolean isMetaFilePresent() {
      * @param metadataMap : Set of XmlMetaData objects
      * @throws IOException
      */
-    public void writeMetaFile(ConcurrentHashMap<String, XmlMetadata> metadataMap) throws IOException {
-        if (this.isMetaFilePresent()) Files.delete(Paths.get(metaFile.getCanonicalPath()));
+    public void writeMetaFile(ConcurrentHashMap<String, XmlMetadata> metadataMap) throws IOException, JAXBException {
+        if (this.isMetaFilePresent())
+            Files.delete(Paths.get(metaFile.getCanonicalPath()));
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/MetaFileUtil.java]: add {}.

**Comment by prestoncarman** [comment]
Can you post a sample vxquery_index.xml file? (you can use the wiki)

**Comment by menakaj** [comment]
Here is the vxquery_index.xml file

On 11 July 2016 at 01:02, Preston Carman notifications@github.com wrote:

> Can you post a sample vxquery_index.xml file? (you can use the wiki)
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/apache/vxquery/pull/82#issuecomment-231606421, or mute
> the thread
> https://github.com/notifications/unsubscribe/AJsco8zZc-8k5GhOdmtG9GoIX8A5_M4dks5qUUi1gaJpZM4JDwel
> .

## 

_Menaka Madushanka Jayawardena_
Faculty of Engineering, http://www.pdn.ac.lk/eng
University of Peradeniyaya.
LinkedIn http://lk.linkedin.com/in/menakajayawardena
TP:- 071 885 1183/ 071 350 5470

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceCollectionRule.java [hunk 13]
```diff
@@ -64,7 +64,8 @@
     @Override
     public boolean rewritePre(Mutable<ILogicalOperator> opRef, IOptimizationContext context) {
         VXQueryOptimizationContext vxqueryContext = (VXQueryOptimizationContext) context;
-        String args[] = getCollectionName(opRef);
+        String args[] = getFunctionalArguments(opRef, BuiltinFunctions.FN_COLLECTION_WITH_TAG_2, BuiltinFunctions
+                .FN_COLLECTION_1);
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceCollectionRule.java:68]: Create final static variable set of FunctionIdentifiers. Then you can use the set method 'contains' in the getFunctionalArguments.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/AbstractCollectionRule.java [hunk 14]
```diff
@@ -58,7 +60,10 @@
      *            Logical operator
      * @return collection name
      */
-    protected String[] getCollectionName(Mutable<ILogicalOperator> opRef) {
+
+    // Add functional identifies.
+    // Rename getFunctionArguments
+    protected String[] getFunctionalArguments(Mutable<ILogicalOperator> opRef, Function... functions) {
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/AbstractCollectionRule.java:66]: Switch second argument to set<FunctionIdentifier>. Then used the contains method for doing the check.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java [hunk 15]
```diff
@@ -0,0 +1,73 @@
...
+    private String index;
+
+    @Override
+    public boolean rewritePre(Mutable<ILogicalOperator> opRef, IOptimizationContext context)
+            throws AlgebricksException {
+        VXQueryOptimizationContext vxqueryContext = (VXQueryOptimizationContext) context;
+        String args[] = getFunctionalArguments(opRef, BuiltinFunctions.FN_BUILD_INDEX_ON_COLLECTION_2, BuiltinFunctions
+                .FN_UPDATE_INDEX_1, BuiltinFunctions.FN_DELETE_INDEX_1);
+
+
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java:32]: Format the file.

**Comment by prestoncarman** [comment]
This pull request will never get merged unless we stop adding new features to the change. Please create a new branch for the parallel changes. The new branch can depend on these changes, but keep it separate. I think we need to limit this pull request to index-update, index-delete and the metadata file.

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceCollectionRule.java (hunk 13 above)
Comment (menakaj) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceCollectionRule.java:68]: It's bit unclear to me creating set of FunctionalIdentifires. Is it a single set of all identifres related to indexing and collection or separate ones.

Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceCollectionRule.java:68]: Something like this:
    private static final Set<FunctionIdentifier> COLLECTIONS = new HashSet<>();
    static {
        COLLECTIONS.add(BuiltinFunctions.FN_COLLECTION_0.getFunctionIdentifier());
        COLLECTIONS.add(BuiltinFunctions.FN_COLLECTION_WITH_TAG_2.getFunctionIdentifier());
    }

File: vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml [hunk 16]
```diff
@@ -100,80 +100,88 @@
...
+    </function>
+
     <!-- fn:collection-from-index($indexfolder  as xs:string?, $elementpath as xs:string?) as  node()* -->
     <function name="fn:collection-from-index">
         <param name="index-folder" type="xs:string?"/>
         <param name="element-path" type="xs:string?"/>
-        <return type="node()*"/>
-        <runtime type="unnesting" class="org.apache.vxquery.runtime.functions.index.CollectionFromIndexUnnestingEvaluatorFactory"/>
+        <return type="xs:boolean?"/>
+         <runtime type="unnesting" class="org.apache.vxquery.runtime.functions.index.CollectionFromIndexUnnestingEvaluatorFactory"/>
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/functions/builtin-functions.xml:160]: Remove extra space. (formatting 4 spaces)

File: vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java [hunk 17]
```diff
@@ -0,0 +1,73 @@
+package org.apache.vxquery.compiler.rewriter.rules;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/compiler/rewriter/rules/IntroduceIndexingRule.java:1]: no license header

File: vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/XmlMetadataCollection.java [hunk 18]
```diff
@@ -0,0 +1,46 @@
+package org.apache.vxquery.runtime.functions.index.updateIndex;
```
Comment (prestoncarman) [inline vxquery-core/src/main/java/org/apache/vxquery/runtime/functions/index/updateIndex/XmlMetadataCollection.java:1]: no license header.

**Pull request closed without merging**